server/src/main/java/org/elasticsearch/action/search/SearchTaskStatus.java

```diff
@@ -0,0 +1,188 @@
+        builder.startObject();
+        {
+            PhaseInfo phaseInfo = currentPhase.get();
+            if (phaseInfo != null) {
+                builder.startObject("phase_running");
+                phaseInfo.toXContent(builder, params);
+                builder.endObject();
+            }
+            builder.startArray("phases_completed");
+            for (PhaseInfo phase : phases) {
+                builder.startObject();
+                phase.toXContent(builder, params);
+                builder.endObject();
+            }
+            builder.endArray();
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    private static class PhaseInfo implements ToXContentFragment, Writeable {
+        private final String name;
+        private final int expectedOps;
+        private final List<ShardInfo> processed;
+
+        PhaseInfo(String name, int expectedOps) {
+            this.name = name;
+            this.expectedOps = expectedOps;
+            this.processed = new CopyOnWriteArrayList<>();
+        }
+
+        PhaseInfo(StreamInput in) throws IOException {
+            this.name = in.readString();
+            this.expectedOps = in.readVInt();
+            this.processed = in.readList(ShardInfo::new);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(name);
+            out.writeVInt(expectedOps);
+            out.writeList(processed);
+        }
+
+        void shardProcessed(ShardId shardId, TaskInfo taskInfo) {
+            processed.add(new ShardInfo(shardId, taskInfo));
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject(name);
+            {
+                builder.field("total_shards", expectedOps);
+                builder.field("processed_shards", processed.size());
+                builder.startArray("processed");
+                for (ShardInfo shardInfo : processed) {
+                    builder.startObject();
+                    {
+                        builder.field("index_name", shardInfo.shardId.getIndex().getName());
+                        builder.field("index_uuid", shardInfo.shardId.getIndex().getUUID());
+                        builder.field("shard_id", shardInfo.shardId.getId());
+                        builder.startObject("task_info");
+                        shardInfo.taskInfo.toXContent(builder, params);
+                        builder.endObject();
+                    }
+                    builder.endObject();
+                }
+                builder.endArray();
+            }
+            builder.endObject();
+            return builder;
+        }
+    }
+
+    private static class ShardInfo implements Writeable {
+        private final ShardId shardId;
+        private final TaskInfo taskInfo;
+
+        ShardInfo(ShardId shardId, TaskInfo taskInfo) {
```

Comment:
TaskInfo can be null if the shard response comes from a node that is in a previous version. Should we take this into account here ?